### ISSUE TITLE
Add JuliaDiffEq Common Interface Bindings

### ISSUE THREAD

**Issue by ChrisRackauckas** [PerezHz/TaylorIntegration.jl#19]
Hey,
  This package has really evolved and looks really cool! I was thinking it would be nice to add common interface bindings. To do this, it would just need to add a dependency on DiffEqBase and implement a `solve` function which translates how to solve an `ODEProblem` to the solver for this package. 

Taking a quick look at the examples, it looks like it just would extract the information of the `ODEProblem` (the function, timespan, and initial condition), and use it to build the

```julia
taylorinteg(kepler_problem, q0, t0, 0.01, order, abs_tol, maxsteps)
```

command. The algorithm could be

```julia
alg = TaylorIntegration(order=SomeDefaultHere)
```

And then

```julia
solve(prob,alg,maxiters=...,abstol=...)
```

would be a solve command that translates the arguments and calls `taylorinteg` appropriately. Would you accept such a PR?

Now, the big issue that I see from the examples is that they're not using a inplace format. Your examples seem to use the ODE.jl `du=f(t,u)` version, but doesn't include support for `f(t,u,du)`. Is that correct? If so, that hampers performance quite a bit, but we can translate any inplace function to the not-inplace form using the same translation as the ODE.jl common interface bindings.

**Comment by ChrisRackauckas** [comment]
How exactly are the timesteps chosen?

**Comment by lbenet** [comment]
Thanks for the comments and the interest! We surely accept PRs!

We have discussed (offline) about adapting our package for the inplace evaluation, just as you have it in `DifferentialEquations.jl`. @PerezHz did some benchmarks and I think we were getting ~%20 improvement. Maybe that, combined with #18, could yield even better results. This is something we are on right now. 

Yet, `TaylorIntegration` is comparing poorly in terms of speed with other methods due to memory allocations, which at the end is due to the way `TaylorSeries.jl` is constructed. One way to go around that is to construct a macro which parses `f(t, u)` (or `f(t, u, du)`) in such a way that we get a function that implements the equations in terms of the Taylor coefficients,; we haven't got anything interesting yet... Also, after some discussions with @dpsanders, we are reconsidering the design of `TaylorSeries.jl`, to make at least such parsing easier.

The time step is chosen to guarantee some sort of convergence condition, which then involves the last few coefficients of the local Taylor expansion. Currently, our implementation considers a measure of the absolute error, which involves `abs_tol` and `order`, but it is easy to consider a relative error. We have different `taylorintegration` methods to allow the user choose the king of output: the user may provide initial and final times for the integration in order to get a dense output (a point for each computed time); alternatively, the user provides a time-range to get the solution evaluated only at certain specified times.

**Comment by ChrisRackauckas** [comment]
>Yet, TaylorIntegration is comparing poorly in terms of speed with other methods due to memory allocations, which at the end is due to the way TaylorSeries.jl is constructed.

While you likely could improve this a lot (but having a cache for the order and making everything in place), I actually wouldn't point to this as the reason. I would highly suspect that even a well-optimized Taylor expansion method will not, "on most problems for standard tolerances", do well. These algorithms only use use evaluations of `f` at one endpoint, so they require the problem to be very smooth, and the speed comes from using a high order. There's related algorithms which do this too: extrapolation methods. What has been found over the years is that Runge-Kutta methods just "generally outperform extrapolation methods", but with a catch. The reason is because high order more as the stepsize decreases, so there is a small enough tolerance such that the higher order makes the extrapolation methods do better.

I would presume the same is true for Taylor expansion methods: Runge-Kutta methods are design to just do well "in the non-asymtopic limit", but they are beaten when you need high accuracy. So I think what you really need to do is find the right test. A smooth chaotic problem like the Lorenz Attractor or the N-body problem, and look at very low tolerances (using ArbFloats or BigFloats, I'm talking 1e-30 or 1e-40 tolerances). For chaotic systems, this kind of tolerance really matter, and I think these methods will be demonstrated as useful for these domains.

FYI, my main reason for adding a common interface binding is to add this algorithm to the DiffEqBenchmarks.jl tests and see what happens.

>One way to go around that is to construct a macro which parses f(t, u) (or f(t, u, du)) in such a way that we get a function that implements the equations in terms of the Taylor coefficients,; we haven't got anything interesting yet... Also, after some discussions with @dpsanders, we are reconsidering the design of TaylorSeries.jl, to make at least such parsing easier.

Would you want to piggy-back onto ParameterizedFunctions.jl? Its macro already calculates a bunch of things like Jacobians and Hessians. It wouldn't be hard to add a flag with generates functions for each of the partial derivatives that you need. If there was a way to just use those then you'd be set.

But I think the more interesting case is when it works outside of any DSL. Currently it already works on general `f`'s, how is that done? Autodifferentiation?

**Comment by lbenet** [comment]
Indeed, the fact that Taylor methods do not simply rely in evaluating `f(t,u)` a few times is related with the amount of memory allocated. The difference with extrapolation methods is that Taylor methods actually construct locally a high-order polynomial solution, valid (locally) for *all* values within some interval of convergence (which is used as the time step). So you actually construct all derivatives of that polynomial. As you point out, Taylor methods are worth if you want high accuracy, which imposes using a large enough order, so that's our prime interest here. Yet, we do get good accuracy without using extended precision!

One of my favorites examples on this is integrating $\dot{x} = x^2$, with $x(0)=1$; you can actually *prove* that the polynomial solution converges to the actual solution, and that the independent variable does not exceed the region where the solution is proved to exist.

Thanks for suggesting taking a look onto `ParameterizedFunctions.jl`; I'll certainly check it.

The way Taylor integration works is that you construct recursively, order by order, the solution. The (n+1)-th Taylor coefficient of the solution is related to the n-th Taylor coefficient of the function `f`. That is the place where `TaylorSeries.jl` enters; `TaylorSeries.jl` is based on automatic (or algorithmic) differentiation. So, one current limitation is the set of functions that `TaylorSeries.jl` covers.

**Comment by PerezHz** [comment]
Hi, @ChrisRackauckas

> This package has really evolved and looks really cool! I was thinking it would be nice to add common interface bindings. To do this, it would just need to add a dependency on DiffEqBase and implement a solve function which translates how to solve an ODEProblem to the solver for this package.

> FYI, my main reason for adding a common interface binding is to add this algorithm to the DiffEqBenchmarks.jl tests and see what happens.

Thank you for your interest! That would be really nice!

> Now, the big issue that I see from the examples is that they're not using a inplace format. Your examples seem to use the ODE.jl du=f(t,u) version, but doesn't include support for f(t,u,du). Is that correct? If so, that hampers performance quite a bit, but we can translate any inplace function to the not-inplace form using the same translation as the ODE.jl common interface bindings.

We have some work being done in this direction in PR #18, which precisely implements in-place evaluation. It is almost ready; we've updated the code, tests, and the example notebooks, feel free to check them out! As always, suggestions and comments are welcome.

**Comment by lbenet** [comment]
I think this was fixed in #44; closing.
